### DISCUSSION TITLE
SDXL controlnet training, sudden convergence

### DISCUSSION THREAD

**Discussion by brycegoh** [huggingface/diffusers#7154] [Q&A]
Hello, I am trying to train a controlnet model using [the provided script](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet_sdxl.py) that is conditioned on semantic segmentations. I am looking for a general gauge for how many steps for [sudden convergence](https://github.com/lllyasviel/ControlNet/blob/main/docs/train.md#more-consideration-sudden-converge-phenomenon-and-gradient-accumulation) on sdxl controlnet especially from the diffusers team who have already trained [sdxl controlnets](https://huggingface.co/collections/diffusers/sdxl-controlnets-64f9c35846f3f06f5abe351f).

For the dataset, I am using the ADE20k dataset (20k image pairs). For captions, I am using blip.

I have completed around 15k steps with learning rate of 1e-5, constant learning rate scheduler, gradient accumulation of 1 and batch size of 6, but still have not noticed any sudden convergence. 

If anyone from the diffuser team reads this, did you notice any patterns in terms of the step range in which convergence happens during the training of the [diffuser's sdxl controlnet](https://huggingface.co/collections/diffusers/sdxl-controlnets-64f9c35846f3f06f5abe351f)? Any advice on the hyper-parameters, mixed-precision config, etc?

Any advice would be appreciated!

**Comment by rphly** [comment]
This is an interesting discussion and will help us as well. Hoping the team can respond to this.

**Comment by sayakpaul** [comment]
Thanks for opening this discussion. Does increasing the learning rate help? 

Additionally, this script might be helpful for training large-scale:
https://github.com/huggingface/diffusers/blob/main/examples/research_projects/controlnet/train_controlnet_webdataset.py

This is what we used for https://huggingface.co/collections/diffusers/sdxl-controlnets-64f9c35846f3f06f5abe351f

**Comment by brycegoh** [comment]
> Thanks for opening this discussion. Does increasing the learning rate help?
> 
> Additionally, this script might be helpful for training large-scale: https://github.com/huggingface/diffusers/blob/main/examples/research_projects/controlnet/train_controlnet_webdataset.py
> 
> This is what we used for https://huggingface.co/collections/diffusers/sdxl-controlnets-64f9c35846f3f06f5abe351f

Thanks for the quick reply Sayak. 

1. May I ask what's the difference between the script you linked and the one in the [example folder](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet_sdxl.py) that makes it more suitable?

2. I used 1e-5 as that's the learning rate used by your team for train the depth controlnet (https://huggingface.co/diffusers/controlnet-depth-sdxl-1.0#hyper-parameters). Therefore, would you advice me to increase the learning rate by either:
 2.1 Literally increasing the learning rate
 2.2 Scale the learning rate by batch number as your team did for the canny controlnet [here](https://huggingface.co/diffusers/controlnet-canny-sdxl-1.0#hyper-parameters)
 2.3 Changing the learning rate scheduler (for example, change constant to cosine). Note that it seems your team used constant learning rate for both canny and depth controlnets.

**Reply by sayakpaul** [reply]
1. It uses a different objective to train. Specifically, it uses the EDM formulation of training as discussed in https://arxiv.org/abs/2206.00364. 
2. I would just start by increasing the learning rate and see what happens. 


We have had folks having success with segmentation images as control conditioning too. Check out https://opensource.googleblog.com/2023/06/controlling-stable-diffusion-with-jax-diffusers-and-cloud-tpus.html.

**Reply by brycegoh** [reply]
Thanks Sayak! Will check them out. Really appreciate your help on this. WIll update this post when I can.

**Reply by brycegoh** [reply]
@sayakpaul do we have any examples of what dataset format does [the script](https://github.com/huggingface/diffusers/blob/main/examples/research_projects/controlnet/train_controlnet_webdataset.py) expect?

**Reply by sayakpaul** [reply]
It’s the `webdataset`.

**Reply by sayakpaul** [reply]
Here's an example of the shards: https://huggingface.co/datasets/laion/conceptual-captions-12m-webdataset. But note that it's missing the conditioning image. 

If you are looking for a script to generate these shards, this could be a nice reference:https://github.com/sayakpaul/instructpix2pix-sdxl/blob/main/upscale_swin2sr/generate_wds_shards.py. 

You might have to repurpose it a bit.

**Reply by brycegoh** [reply]
Hi Sayak, thanks for the examples and script.

Would like your help on an issue I am facing regarding loading the training dataset using webdataset. 

This is my dataset, https://huggingface.co/datasets/helloHearthXHome/scene-150-shards

This is the command (Note that i am testing with only 1 .tar file):
```
!accelerate launch diffusers/examples/research_projects/controlnet/train_controlnet_webdataset.py \
 --pretrained_model_name_or_path=stabilityai/stable-diffusion-xl-base-1.0 \
 --pretrained_vae_model_name_or_path=madebyollin/sdxl-vae-fp16-fix \
 --train_shards_path_or_url=/data/00000.tar \
 --learning_rate=1e-5 \
 --lr_scheduler=constant \
 --max_train_steps=10000 \
 --train_batch_size=2 \
 --max_train_samples=1000 \
 --gradient_accumulation_steps=1 \
 --checkpointing_steps=500 \
 --validation_image=control_img.png \
 --validation_prompt="validation prompt" \
 --validation_steps=50 \
 --hub_model_id=HUB_ID \
 --tracker_project_name=SDXL_CONTROLNET \
 --resolution=1024 \
 --report_to="wandb" \
 --enable_xformers_memory_efficient_attention \
 --mixed_precision=fp16 \
 --dataloader_num_workers=4 \
 --push_to_hub
```

Modification to the script is only done to `Text2ImageDataset` `__init__` method:
<img width="510" alt="image" src="https://github.com/huggingface/diffusers/assets/16372429/00f8ae4a-3cc9-49fe-8cb3-944d377acc99">

I tried running the training script but it seems to get stuck without any errors or warnings after loading the progress bar. Is my dataset format wrong? Any advice would be appreciated!

**Reply by sayakpaul** [reply]
I would recommend taking a step back and just trying to step through the dataloader in a separate script with multiple print statements thrown in between. This way, you will be able to point out the exact codepath that needs fixing.

**Reply by brycegoh** [reply]
Thanks Sayak, I did try that before. Let me explain it

I firstly added print statements to the `group_by_keys_nothrow` function that yields samples as such:
<img width="689" alt="image" src="https://github.com/huggingface/diffusers/assets/16372429/57bdcbc9-0fe9-467d-b57f-06f0161766fa">

The stdouts are like `0 jpg` meaning `prefix = 0` and `suffix = jpg`

The prints just keeps looping through even though for each tar file I set the maximum number of samples to be 1000 meaning a total of 4000 files given that each sample has image, control_image, original_size and text.

std out example:
<img width="114" alt="image" src="https://github.com/huggingface/diffusers/assets/16372429/0d1cdfb5-36f4-446f-a519-38d51cdca6d5">

My code to get the first batch:
```python
dataset = Text2ImageDataset(
        train_shards_path_or_url="00000.tar",
        eval_shards_path_or_url=None,
        num_train_examples=1000,
        per_gpu_batch_size=1,
        global_batch_size=1,
        num_workers=1,
        resolution=1024,
        center_crop=False,
        random_flip=False,
        shuffle_buffer_size=1000,
        pin_memory=True,
        persistent_workers=True,
        control_type="canny",
        feature_extractor=None,
    )
train_dataloader = dataset.train_dataloader

print("start dataset debugging")
try:
    first_batch = next(iter(train_dataloader))
    print("Successfully fetched a batch:", first_batch)
except StopIteration:
    print("The dataloader did not yield any batches.")
except Exception as e:
    print(f"An error occurred: {e}")
```

I am unsure if this is expected


## Alternative
Is it fair to just change the noise scheduler on the [the provided script](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet_sdxl.py) to the `EulerDiscreteScheduler` instead of `DDPMScheduler`? Or are you aware if there are more in-depth changes?

**Reply by sayakpaul** [reply]
Are you able to print the shapes of the images and other columns? 

I will also recommend using a simpler data input pipeline to see if you are able to successfully load from the tars and can stream from it. 

> Is it fair to just change the noise scheduler on the [the provided script](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet_sdxl.py) to the EulerDiscreteScheduler instead of DDPMScheduler? Or are you aware if there are more in-depth changes?

The script uses EDM formulation. So, DDPM won't work.

**Reply by brycegoh** [reply]
Ok, we have 2 different scripts here:

A) The one provided with documentation (https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet_sdxl.py)
B) The one that your diffuser team used (https://github.com/huggingface/diffusers/blob/main/examples/research_projects/controlnet/train_controlnet_webdataset.py)

B uses the `EulerDiscreteScheduler` and also the EDM formulation while A uses `DDPMScheduler`.

Therefore, instead of adjusting my dataset to webdataset format, could I intergrate the EDM formulation portion of the script B into script A? 

If that's possible, would appreciate if you could point me in the right direction such as 
1. where are the essential portions of the EDM formulation in script B that differs from script A?
2. Is it purely just EDM formulation that differs?

-------------

I am asking point 2 as well because comparing the code, there are slight differences such as:
```python

 ### Script A
latents = vae.encode(pixel_values).latent_dist.sample() 
latents = torch.cat(latents, dim=0)


### Script B
# encode pixel values with batch size of at most 8.                                            
latents = []                                                                                                                 
for i in range(0, pixel_values.shape[0], 8):                                                          
    latents.append(vae.encode(pixel_values[i : i + 8]).latent_dist.sample())     
latents = torch.cat(latents, dim=0)                                                                        
```

Apologies as I am not familiar with the theory behind this.

**Reply by brycegoh** [reply]
@sayakpaul I tried referencing your recent [PR](https://github.com/huggingface/diffusers/pull/7126) that was included in the latest [v0.27.0 release](https://github.com/huggingface/diffusers/releases/tag/v0.27.0), https://github.com/huggingface/diffusers/pull/7126

My modifications can be referenced here, https://github.com/brycegoh/diffusers/pull/1
I made a local PR because it is evidently not ready for a PR on the official diffusers repo without the additional test cases and README modifications.

Apart from EDM style training modifications, do we need to add the `crop_top_lefts` coordinates and original sizes to each batch. I saw that the dataset loader on your PR keeps track of the coordinates instead of relying on the command inputs of default 0. I am not sure if that is needed. However, it does seem useless as we already resize and then do a centre crop so top left is always 0,0?

Would appreciate some validation, checks and guidance on my modifications.
